core/src/main/kotlin/org/jetbrains/kotlinx/dataframe/impl/api/schema.kt

```diff
@@ -0,0 +1,54 @@
+package org.jetbrains.kotlinx.dataframe.impl.api
+
+import org.jetbrains.kotlinx.dataframe.columns.ColumnPath
+import org.jetbrains.kotlinx.dataframe.impl.schema.DataFrameSchemaImpl
+import org.jetbrains.kotlinx.dataframe.impl.schema.getSchema
+import org.jetbrains.kotlinx.dataframe.schema.ColumnSchema
+import org.jetbrains.kotlinx.dataframe.schema.DataFrameSchema
+import kotlin.reflect.KClass
+
+@PublishedApi
+internal fun compileTimeSchemaImpl(runtimeSchema: DataFrameSchema, klass: KClass<*>): DataFrameSchema {
+    val compileSchema = getSchema(klass)
+    val root = ColumnPath(emptyList())
+    val order = mutableMapOf<ColumnPath, Int>()
+    runtimeSchema.putColumnsOrder(order, path = root)
+    return compileSchema.sorted(order, path = root)
+}
+
+internal fun DataFrameSchema.putColumnsOrder(order: MutableMap<ColumnPath, Int>, path: ColumnPath) {
+    columns.entries.forEachIndexed { i, (name, column) ->
+        val columnPath = path + name
+        order[columnPath] = i
+        when (column) {
+            is ColumnSchema.Frame -> {
+                column.schema.putColumnsOrder(order, columnPath)
+            }
+
+            is ColumnSchema.Group -> {
+                column.schema.putColumnsOrder(order, columnPath)
+            }
+        }
+    }
+}
+
+internal fun DataFrameSchema.sorted(order: Map<ColumnPath, Int>, path: ColumnPath): DataFrameSchema {
```

Comment:
here it's okay to have it as a receiver :) I'd only name it "sortedBy(order)" to be more expressive.